java-checks/src/main/java/org/sonar/java/checks/security/DebugFeatureEnabledCheck.java

```diff
@@ -67,4 +109,12 @@ private static AssignmentExpressionTree getDebugArgument(ExpressionTree expressi
     return null;
   }
 
+  private boolean enclosingClassExtendsThrowable() {
+    return enclosingClass.peek() != null && enclosingClass.peek().type().isSubtypeOf("java.lang.Throwable");
+  }
+
+  private static boolean calledOnTypeInheritedFromThrowable(ExpressionTree tree) {
+    return tree.symbolType().isSubtypeOf("java.lang.Throwable");
+  }
+
```

Comment:
With a MethodMatcher:

- the above two methods can be removed
- `enclosingClass` field can be removed
- nodesToVisit() on  Tree.Kind.CLASS can be removed
- leaveNode() can be removed

:sunglasses:

Reply:
The `enclosingClassExtendsThrowable` method (along witht the `enclosingClass` field) is there to prevent issues from being reported when calling `printWithStackTrace` inside classes extending `Throwable`.

As far as I can tell there's no way to say "Don't match when inside a class inheriting from X" with `MethodMatchers`, so that part of the code seems to still be necessary.

Reply:
You are right, we need to keep `enclosingClass`, I didn't understand it correctly previously.